File: apps/ios/Exponential/UI/Session/AgentUsageWindowPrefs.swift
import Foundation

/// EXP-484: which rate-limit window this phone shows for an agent's usage bar.
///
/// Deliberately LOCAL and per-client — the desktop keeps its own in
/// `coding::Settings`, the web its own in localStorage. It is a viewing
/// preference, not machine state, so it never rides the device wire.
/// An unremembered agent falls back to the fullest window
/// (`AgentUsagePresentation.selectWindow`).
enum AgentUsageWindowPrefs {
    private static func key(for agent: String) -> String { "agentUsageWindow.\(agent)" }

    /// The remembered window key, or nil when this phone has never picked one
    /// for [agent] (or picked one the machine no longer reports).
    static func read(agent: String) -> String? {
        let stored = UserDefaults.standard.string(forKey: key(for: agent))
        guard let stored, !stored.isEmpty else { return nil }
        return stored
    }

    static func remember(agent: String, key windowKey: String) {
        UserDefaults.standard.set(windowKey, forKey: key(for: agent))
    }
}

File: apps/android/app/src/main/java/com/exponential/app/data/AgentUsageWindowPrefs.kt
package com.exponential.app.data

import com.exponential.app.data.auth.SecureStore
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * EXP-484: which usage window a user last picked for an agent — a purely
 * LOCAL preference (the desktop, web and iOS each keep their own; nothing is
 * synced), so a collapsed bar keeps showing the window that person cares
 * about instead of whichever one happens to be busiest.
 *
 * Keyed by agent, not by device: the same person watching the same agent on
 * two machines wants the same window. An unknown/no-longer-reported key just
 * falls back to the busiest window (see [AgentUsagePresentation.selectWindow]).
 *
 * SecureStore isn't observable, so — the TeamSelection.rememberLastBoard idiom
 * — a version counter lets reactive readers re-read after every write.
 */
@Singleton
class AgentUsageWindowPrefs @Inject constructor(
    private val secureStore: SecureStore,
) {
    private val _version = MutableStateFlow(0)
    val version: StateFlow<Int> = _version.asStateFlow()

    fun read(agent: String): String? =
        if (agent.isBlank()) null else secureStore.get(prefKey(agent))

    fun remember(agent: String, key: String) {
        if (agent.isBlank() || key.isBlank()) return
        secureStore.set(prefKey(agent), key)
        _version.value += 1
    }

    private fun prefKey(agent: String) = "agent_usage_window_$agent"
}
